collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java

```diff
@@ -0,0 +1,114 @@
+package org.odk.collect.android.database;
+
+import android.content.ContentValues;
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import org.jetbrains.annotations.NotNull;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.odk.collect.android.database.DatabaseConstants.FORMS_TABLE_NAME;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.AUTO_DELETE;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.AUTO_SEND;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.BASE64_RSA_PUBLIC_KEY;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.DATE;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.DESCRIPTION;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.DISPLAY_NAME;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.FORM_FILE_PATH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.FORM_MEDIA_PATH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.GEOMETRY_XPATH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.JRCACHE_FILE_PATH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.JR_FORM_ID;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.JR_VERSION;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.LANGUAGE;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.MD5_HASH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.SUBMISSION_URI;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns._ID;
+
+@RunWith(AndroidJUnit4.class)
+public class FormDatabaseMigratorTest {
```

Comment:
There were tests for the database helper checking migrations, but they were out of date and `Ignore`d. Interested to know if people think we need more coverage here.

Reply:
I know you've been trying to kill those tests for a long time. 😂 They were added after we had several significant bugs related to `onDowngrade`. I think the structure and test are good but that it would be good to also downgrade from a fictitious high version with an extra and missing column to avoid regressions there. We can assume we wouldn't touch `onDowngrade` anymore but it's hard to reason about and tempting to try to improve.

Reply:
haha yeah totally happy to add more. The old tests were failing and needed rewritten anyway. I'll add a couple of downgrade tests (one with extra and one with missing columns).